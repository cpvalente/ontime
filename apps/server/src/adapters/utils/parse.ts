/**
 * @description Creates a nested object with keys from an array and assigns the `value` to the last key
 * @param {array} path - array to be nested
 * @param {string} value - value to assign
 * @returns {object | string | null} nested object or null if no object was created
 */
export const integrationPayloadFromPath = (path: string[], value?: unknown): object | string | null => {
  if (path.length === 1) {
    const key = path[0];
    return value === undefined ? key : { [key]: value };
  }

  const parsedValue = value === undefined ? path.at(-1) : value;
  const shortenedPath = value === undefined ? path.slice(0, -1) : path;

  const obj = shortenedPath.reduceRight((result, key) => ({ [key]: result }), parsedValue);

  return typeof obj === 'object' ? obj : null;
};
