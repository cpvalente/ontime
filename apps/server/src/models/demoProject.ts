import { DatabaseModel, EndAction, SupportedEvent, TimeStrategy, TimerType } from 'ontime-types';

export const demoDb: DatabaseModel = {
  rundown: [
    {
      type: SupportedEvent.Event,
      id: '32d31',
      cue: 'SF1.01',
      title: 'Albania',
      note: 'SF1.01',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 36000000,
      timeEnd: 37200000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Sekret',
        artist: 'Ronela Hajati',
      },
    },
    {
      type: SupportedEvent.Event,
      id: '21cd2',
      cue: 'SF1.02',
      title: 'Latvia',
      note: 'SF1.02',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 37500000,
      timeEnd: 38700000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Eat Your Salad',
        artist: 'Citi Zeni',
      },
    },
    {
      type: SupportedEvent.Event,
      id: '0b371',
      cue: 'SF1.03',
      title: 'Lithuania',
      note: 'SF1.03',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 39000000,
      timeEnd: 40200000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Sentimentai',
        artist: 'Monika Liu',
      },
    },
    {
      type: SupportedEvent.Event,
      id: '3cd28',
      cue: 'SF1.04',
      title: 'Switzerland',
      note: 'SF1.04',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 40500000,
      timeEnd: 41700000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Boys Do Cry',
        artist: 'Marius Bear',
      },
    },
    {
      type: SupportedEvent.Event,
      id: 'e457f',
      cue: 'SF1.05',
      title: 'Slovenia',
      note: 'SF1.05',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 42000000,
      timeEnd: 43200000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Disko',
        artist: 'LPS',
      },
    },
    {
      type: SupportedEvent.Block,
      id: '01e85',
      title: 'Lunch break',
    },
    {
      type: SupportedEvent.Event,
      id: '1c420',
      cue: 'SF1.06',
      title: 'Ukraine',
      note: 'SF1.06',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 47100000,
      timeEnd: 48300000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Stefania',
        artist: 'Kalush Orchestra',
      },
    },
    {
      type: SupportedEvent.Event,
      id: 'b7737',
      cue: 'SF1.07',
      title: 'Bulgaria',
      note: 'SF1.07',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 48600000,
      timeEnd: 49800000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Intention',
        artist: 'Intelligent Music Project',
      },
    },
    {
      type: SupportedEvent.Event,
      id: 'd3a80',
      cue: 'SF1.08',
      title: 'Netherlands',
      note: 'SF1.08',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 50100000,
      timeEnd: 51300000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'De Diepte',
        artist: 'S10',
      },
    },
    {
      type: SupportedEvent.Event,
      id: '8276c',
      cue: 'SF1.09',
      title: 'Moldova',
      note: 'SF1.09',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 51600000,
      timeEnd: 52800000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Trenuletul',
        artist: 'Zdob si Zdub',
      },
    },
    {
      type: SupportedEvent.Event,
      id: '2340b',
      cue: 'SF1.10',
      title: 'Portugal',
      note: 'SF1.10',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 53100000,
      timeEnd: 54300000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Saudade Saudade',
        artist: 'Maro',
      },
    },
    {
      type: SupportedEvent.Block,
      id: 'cb90b',
      title: 'Afternoon break',
    },
    {
      type: SupportedEvent.Event,
      id: '503c4',
      cue: 'SF1.11',
      title: 'Croatia',
      note: 'SF1.11',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 56100000,
      timeEnd: 57300000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Guilty Pleasure',
        artist: 'Mia Dimsic',
      },
    },
    {
      type: SupportedEvent.Event,
      id: '5e965',
      cue: 'SF1.12',
      title: 'Denmark',
      note: 'SF1.12',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,

      timeStart: 57600000,
      timeEnd: 58800000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'The Show',
        artist: 'Reddi',
      },
    },
    {
      type: SupportedEvent.Event,
      id: 'bab4a',
      cue: 'SF1.13',
      title: 'Austria',
      note: 'SF1.13',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 59100000,
      timeEnd: 60300000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Halo',
        artist: 'LUM!X & Pia Maria',
      },
    },
    {
      type: SupportedEvent.Event,
      id: 'd3eb1',
      cue: 'SF1.14',
      title: 'Greece',
      note: 'SF1.14',
      endAction: EndAction.None,
      timerType: TimerType.CountDown,
      linkStart: null,
      timeStrategy: TimeStrategy.LockEnd,
      timeStart: 60600000,
      timeEnd: 61800000,
      duration: 1200000,
      isPublic: true,
      skip: false,
      colour: '',
      revision: 0,
      timeWarning: 500000,
      timeDanger: 100000,
      custom: {
        song: 'Die Together',
        artist: 'Amanda Tenfjord',
      },
    },
  ],
  project: {
    title: 'Eurovision Song Contest',
    description: 'Turin 2022',
    publicUrl: 'www.getontime.no',
    publicInfo: 'Rehearsal Schedule - Turin 2022',
    backstageUrl: 'www.github.com/cpvalente/ontime',
    backstageInfo: 'Rehearsal Schedule - Turin 2022\nAll performers to wear full costumes for 1st rehearsal',
  },
  settings: {
    app: 'ontime',
    version: '3.3.2',
    serverPort: 4001,
    editorKey: null,
    operatorKey: null,
    timeFormat: '24',
    language: 'en',
  },
  viewSettings: {
    dangerColor: '#ED3333',
    endMessage: '',
    freezeEnd: false,
    normalColor: '#ffffffcc',
    overrideStyles: false,
    warningColor: '#FFAB33',
  },
  customFields: {
    song: {
      label: 'Song',
      type: 'string',
      colour: '#339E4E',
    },
    artist: {
      label: 'Artist',
      type: 'string',
      colour: '#3E75E8',
    },
  },
  urlPresets: [
    {
      enabled: true,
      alias: 'test',
      pathAndParams: 'lower?bg=ff2&text=f00&size=0.6&transition=5',
    },
  ],
  osc: {
    portIn: 8888,
    portOut: 9999,
    targetIP: '127.0.0.1',
    enabledIn: true,
    enabledOut: false,
    subscriptions: [],
  },
  http: {
    enabledOut: false,
    subscriptions: [],
  },
};
