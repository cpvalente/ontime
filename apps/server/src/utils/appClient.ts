/**
 * A simple HTTP client utility to abstract fetch and mimic got's behavior for JSON POST requests.
 */

interface PostJsonOptions {
  json: Record<string, unknown>;
  // Add other options here if needed in the future, like headers
}

export async function postJson<T = any>(url: string, options: PostJsonOptions): Promise<T> {
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        // Add any other default headers if necessary
      },
      body: JSON.stringify(options.json),
    });

    if (!response.ok) {
      let errorBodyText = '';
      try {
        // Try to parse as JSON first, as it might be structured error info
        const errorBodyJson = await response.json();
        errorBodyText = JSON.stringify(errorBodyJson);
      } catch (e) {
        // If not JSON, try to get it as plain text.
        try {
          errorBodyText = await response.text();
        } catch (textError) {
          // If reading as text also fails, note that.
          errorBodyText = `(Failed to read error response body: ${
            textError instanceof Error ? textError.message : String(textError)
          })`;
        }
      }
      throw new Error(
        `Request to ${url} failed with status ${response.status} (${response.statusText}). Response body: ${errorBodyText}`,
      );
    }

    // got by default parses JSON response
    return response.json() as Promise<T>;
  } catch (error) {
    if (error instanceof Error) {
      // If it's already an Error (e.g. the one we threw above, or a network error from fetch),
      // re-throw it. We could also choose to wrap it if we wanted to add more context,
      // but for simplicity, re-throwing is fine.
      throw error;
    }
    // For any other type of thrown value, convert to string and wrap in a new Error.
    throw new Error(`Network or unexpected error during request to ${url}: ${String(error)}`);
  }
}
