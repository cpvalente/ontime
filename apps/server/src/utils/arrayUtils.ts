/**
 * Inserts an item in an array at a given index
 * @param index
 * @param item
 * @param array
 */
export function insertAtIndex<T>(index: number, item: T, array: T[]): T[] {
  const modifiedArray = [...array];

  // Insert at beginning
  if (index === 0) {
    modifiedArray.unshift(item);
  }

  // insert at end
  else if (index >= modifiedArray.length) {
    modifiedArray.push(item);
  }

  // insert in the middle
  else {
    modifiedArray.splice(index, 0, item);
  }

  return modifiedArray;
}

/**
 * Deletes array element at a given index
 * @param index
 * @param array
 */
export function deleteAtIndex<T>(index: number, array: T[]) {
  return array.filter((_, i) => i !== index);
}

export function reorderArray<T>(array: T[], fromIndex: number, toIndex: number) {
  if (fromIndex === toIndex) {
    return array; // No change needed, return the original array
  }

  const modifiedArray = [...array];

  // delete in from
  const [reorderedItem] = modifiedArray.splice(fromIndex, 1);

  // reinsert item at to
  modifiedArray.splice(toIndex, 0, reorderedItem);
  return modifiedArray;
}
