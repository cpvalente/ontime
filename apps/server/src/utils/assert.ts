export function isString(value: unknown): asserts value is string {
  if (typeof value !== 'string') {
    throw new Error(`Unexpected payload type: ${String(value)}`);
  }
}

export function isDefined<T>(value: T | undefined): asserts value is T {
  if (value === undefined) {
    throw new Error('Payload not found');
  }
}

export function isNumber(value: unknown): asserts value is number {
  if (typeof value !== 'number') {
    throw new Error(`Unexpected payload type: ${String(value)}`);
  }
}

export function isObject(value: unknown): asserts value is object {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error(`Unexpected payload type: ${String(value)}`);
  }
}
