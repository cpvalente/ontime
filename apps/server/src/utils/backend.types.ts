export type OntimeError = { message: string };
