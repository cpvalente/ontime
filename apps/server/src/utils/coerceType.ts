/**
 * @description Converts a value to a number if possible, throws otherwise
 * @param {unknown} value - Value to be converted to a string.
 * @returns {string} - The converted value as a string.
 * @throws {Error} Throws an error if the value is null or undefined.
 */
export function coerceString(value: unknown): string {
  if (value == null) {
    throw new Error('Invalid value received');
  }
  return String(value);
}

/**
 * @description Converts a value to a number if possible, throws otherwise
 * @param {unknown} value - Value to be converted to a boolean.
 * @returns {boolean} - The converted value as a boolean.
 * @throws {Error} Throws an error if the value is null or undefined.
 */
export function coerceBoolean(value: unknown): boolean {
  if (value == null) {
    throw new Error('Invalid value received');
  }
  return Boolean(value);
}

/**
 * @description Converts a value to a number if possible, throws otherwise
 * @param {unknown} value - Value to be converted to a number.
 * @returns {number} - The converted value as a number.
 * @throws {Error} Throws an error if the value is null, undefined or not a valid number.
 */
export function coerceNumber(value: unknown): number {
  if (value == null) {
    throw new Error('Invalid value received');
  }
  const parsedValue = Number(value);
  if (isNaN(parsedValue)) {
    throw new Error('Invalid value received');
  }
  return parsedValue;
}
