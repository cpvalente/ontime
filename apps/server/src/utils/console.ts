/* eslint-disable no-console */

/**
 * Utility function to log success messages
 */
export function consoleSuccess(message: string) {
  console.log(inGreen(message));
}

/**
 * Utility function to log error messages
 */
export function consoleError(message: string) {
  console.error(inRed(message));
}

/**
 * Utility function to log highlighted messages
 */
export function consoleHighlight(message: string) {
  console.log(inCyan(message));
}

/**
 * Utility function to log messages with subdued appearance
 */
export function consoleSubdued(message: string) {
  console.log(inGray(message));
}

/**
 * Utility function for console, styles text in green
 */
function inGreen(message: string): string {
  return `\x1b[32m${message}\x1b[0m`;
}

/**
 * Utility function for console, styles text in red
 */
function inRed(message: string): string {
  return `\x1b[31m${message}\x1b[0m`;
}

/**
 * Utility function for console, styles text in cyan
 */
function inCyan(message: string): string {
  return `\x1b[96m${message}\x1b[0m`;
}

/**
 * Utility function for console, styles text in gray
 */
function inGray(message: string): string {
  return `\x1b[2m${message}\x1b[0m`;
}
