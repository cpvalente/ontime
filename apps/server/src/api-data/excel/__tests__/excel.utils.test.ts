import { CustomFields } from 'ontime-types';
import { defaultImportMap, ImportMap } from 'ontime-utils';

import { demoDb } from '../../../models/demoProject.js';

import { getCustomFieldData, rundownToTabular } from '../excel.utils.js';

describe('getCustomFieldData()', () => {
  it('generates a list of keys from the given import map', () => {
    const importMap: ImportMap = {
      worksheet: 'event schedule',
      timeStart: 'time start',
      linkStart: 'link start',
      timeEnd: 'time end',
      duration: 'duration',
      flag: 'flag',
      cue: 'cue',
      title: 'title',
      countToEnd: 'count to end',
      skip: 'skip',
      note: 'notes',
      colour: 'colour',
      endAction: 'end action',
      timerType: 'timer type',
      timeWarning: 'warning time',
      timeDanger: 'danger time',
      custom: {
        lighting: 'lx',
        sound: 'sound',
        video: 'av',
      },
      id: 'id',
    };

    const result = getCustomFieldData(importMap, {});
    expect(result.mergedCustomFields).toStrictEqual({
      lighting: {
        type: 'text',
        colour: '',
        label: 'lighting',
      },
      sound: {
        type: 'text',
        colour: '',
        label: 'sound',
      },
      video: {
        type: 'text',
        colour: '',
        label: 'video',
      },
    });

    // it is an inverted record of <importKey, ontimeKey>
    expect(result.customFieldImportKeys).toStrictEqual({
      lx: 'lighting',
      sound: 'sound',
      av: 'video',
    });
  });

  it('keeps colour information from existing fields', () => {
    const importMap: ImportMap = {
      worksheet: 'event schedule',
      timeStart: 'time start',
      linkStart: 'link start',
      timeEnd: 'time end',
      duration: 'duration',
      flag: 'flag',
      cue: 'cue',
      title: 'title',
      countToEnd: 'count to end',
      skip: 'skip',
      note: 'notes',
      colour: 'colour',
      endAction: 'end action',
      timerType: 'timer type',
      timeWarning: 'warning time',
      timeDanger: 'danger time',
      custom: {
        lighting: 'lx',
        sound: 'sound',
        video: 'av',
        'ontime key': 'excel label',
      },
      id: 'id',
    };

    const existingCustomFields: CustomFields = {
      lighting: { label: 'lighting', type: 'text', colour: 'red' },
      sound: { label: 'sound', type: 'text', colour: 'green' },
      ontime_key: { label: 'ontime key', type: 'text', colour: 'blue' },
    };

    const result = getCustomFieldData(importMap, existingCustomFields);
    expect(result.mergedCustomFields).toStrictEqual({
      lighting: {
        type: 'text',
        colour: 'red',
        label: 'lighting',
      },
      sound: {
        type: 'text',
        colour: 'green',
        label: 'sound',
      },
      video: {
        type: 'text',
        colour: '',
        label: 'video',
      },
      ontime_key: {
        type: 'text',
        colour: 'blue',
        label: 'ontime key',
      },
    });

    // it is an inverted record of <importKey, ontimeKey>
    expect(result.customFieldImportKeys).toStrictEqual({
      lx: 'lighting',
      sound: 'sound',
      av: 'video',
      'excel label': 'ontime_key',
    });
  });

  it('lowercases the keys in the import map', () => {
    const importMap: ImportMap = {
      ...defaultImportMap,
      custom: {
        Lighting: 'Lx',
        Sound: 'sound',
        video: 'av',
      },
    };

    const result = getCustomFieldData(importMap, {});
    expect(result.mergedCustomFields).toStrictEqual({
      Lighting: {
        type: 'text',
        colour: '',
        label: 'Lighting',
      },
      Sound: {
        type: 'text',
        colour: '',
        label: 'Sound',
      },
      video: {
        type: 'text',
        colour: '',
        label: 'video',
      },
    });

    // notice that the keys excel keys are lowercased
    expect(result.customFieldImportKeys).toStrictEqual({
      lx: 'Lighting',
      sound: 'Sound',
      av: 'video',
    });
  });
});

describe('rundownToTabular()', () => {
  it('returns an array of arrays describing a rundown', () => {
    const testRundown = demoDb.rundowns['default'];
    const testCustomFields = demoDb.customFields;

    const result = rundownToTabular(testRundown, testCustomFields);
    expect(result).toMatchSnapshot();
  });
});
