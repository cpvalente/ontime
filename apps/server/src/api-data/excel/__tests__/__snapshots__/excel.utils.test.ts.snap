// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rundownToTabular() > returns an array of arrays describing a rundown 1`] = `
[
  [
    "ID",
    "Flag",
    "Cue",
    "Title",
    "Colour",
    "Time Start",
    "Time End",
    "Duration",
    "Link Start",
    "Count to end",
    "Note",
    "Timer Type",
    "End Action",
    "Warning Time",
    "Danger Time",
    "Skip",
    "Custom Field",
    "Images",
  ],
  [
    "e2163f",
    "",
    "Demo",
    "Clear all, or Create New Project to start fresh",
    "#9d9d9d",
    "",
    "",
    "",
    "",
    "",
    "Go to Settings cog wheel top left of Editor, Project > Create",
    "milestone",
    "",
    "",
    "",
    "",
    "Create Custom Fields to store specific text information or images.

Go to Settings > Project data > Custom fields",
    "https://www.getontime.no/images/icons/ontime-logo.png",
  ],
  [
    "7eaf99",
    "",
    "",
    "Morning Sessions",
    "#339E4E",
    "10:00:00",
    "12:00:00",
    "",
    "",
    "",
    "",
    "group",
    "",
    "",
    "",
    "",
    "",
    "",
  ],
  [
    "9bf60f",
    "",
    "1",
    "Pre-show Countdown",
    "#77C785",
    "10:00:00",
    "11:00:00",
    "01:00:00",
    "",
    "",
    "Music plays, holding slide on screens",
    "count-down",
    "",
    "00:10:00",
    "00:05:00",
    "",
    "Put additional info here",
    "",
  ],
  [
    "bf71a2",
    "",
    "Standby",
    "10:45 - Presenters ready side stage",
    "#A790F5",
    "",
    "",
    "",
    "",
    "",
    "",
    "milestone",
    "",
    "",
    "",
    "",
    "",
    "",
  ],
  [
    "c2697f",
    "",
    "1.1",
    "Welcome",
    "#FFCC78",
    "11:00:00",
    "11:10:00",
    "00:10:00",
    "TRUE",
    "",
    "Emma Thompson",
    "count-down",
    "",
    "00:02:00",
    "00:01:00",
    "",
    "",
    "",
  ],
  [
    "fa593e",
    "TRUE",
    "1.2",
    "Session 1",
    "#77C785",
    "11:10:00",
    "12:00:00",
    "00:50:00",
    "TRUE",
    "",
    "Liam Carter, Sophia Patel + PowerPoint",
    "count-down",
    "",
    "00:02:00",
    "00:01:00",
    "",
    "",
    "",
  ],
  [
    "a8b0b3",
    "",
    "House",
    "11:30 - House staff setup lunch in lobby",
    "#A790F5",
    "",
    "",
    "",
    "",
    "",
    "",
    "milestone",
    "",
    "",
    "",
    "",
    "",
    "",
  ],
  [
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "group-end",
    "",
    "",
    "",
    "",
  ],
  [
    "f60403",
    "",
    "",
    "Lunch",
    "#3E75E8",
    "12:00:00",
    "13:00:00",
    "",
    "",
    "",
    "",
    "group",
    "",
    "",
    "",
    "",
    "",
    "",
  ],
  [
    "0aaa7d",
    "",
    "2.1",
    "Lunch / Countdown to next session",
    "#779BE7",
    "12:00:00",
    "13:00:00",
    "01:00:00",
    "TRUE",
    "",
    "Buffet in lobby
Music plays, holding slide on screens",
    "count-down",
    "",
    "00:02:00",
    "00:01:00",
    "",
    "",
    "",
  ],
  [
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "group-end",
    "",
    "",
    "",
    "",
  ],
  [
    "6b0edb",
    "",
    "",
    "Afternoon Sessions",
    "#339E4E",
    "13:00:00",
    "14:00:00",
    "",
    "",
    "",
    "",
    "group",
    "",
    "",
    "",
    "",
    "",
    "",
  ],
  [
    "02afca",
    "",
    "Standby",
    "12:45 - Presenters ready side stage",
    "#A790F5",
    "",
    "",
    "",
    "",
    "",
    "",
    "milestone",
    "",
    "",
    "",
    "",
    "",
    "",
  ],
  [
    "75ce86",
    "",
    "3.1",
    "Session 2",
    "#77C785",
    "13:00:00",
    "13:50:00",
    "00:50:00",
    "TRUE",
    "",
    "Ethan Brooks + PowerPoint + Video playback",
    "count-down",
    "",
    "00:02:00",
    "00:01:00",
    "",
    "",
    "",
  ],
  [
    "e10ed9",
    "",
    "3.2",
    "Wrap up",
    "#FFCC78",
    "13:50:00",
    "14:00:00",
    "00:10:00",
    "",
    "",
    "Lucas Bennett",
    "count-down",
    "",
    "00:02:00",
    "00:01:00",
    "",
    "",
    "",
  ],
  [
    "07df89",
    "",
    "Strike",
    "14:00 - AV & Room strike",
    "#A790F5",
    "",
    "",
    "",
    "",
    "",
    "",
    "milestone",
    "",
    "",
    "",
    "",
    "",
    "",
  ],
  [
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "group-end",
    "",
    "",
    "",
    "",
  ],
]
`;
